src/reconcile/rec_visibility.c

```diff
@@ -586,35 +587,70 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
+         */
+        WT_ASSERT(session, upd_select->upd != NULL);
+        if (upd_select->upd->type == WT_UPDATE_TOMBSTONE) {
+            upd_select->stop_ts = upd_select->upd->start_ts;
+            upd_select->stop_txn = upd_select->upd->txnid;
+            upd_select->upd = upd_select->upd->next;
+        }
+        if (upd_select->upd != NULL) {
+            /* The beginning of the validity window is the selected update's time pair. */
             upd_select->durable_ts = upd_select->start_ts = upd_select->upd->start_ts;
-        if (upd_select->upd->txnid != WT_TXN_NONE)
             upd_select->start_txn = upd_select->upd->txnid;
+        } else {
+            /*
+             * The only time we don't get the unpacked value is when we insert a key in which case
+             * we can't have a tombstone.
+             */
+            WT_ASSERT(session, vpack != NULL);
+            /*
+             * It's possible to have a tombstone as the only update in the update list. If we
+             * reconciled before with only a single update and then read the page back into cache,
+             * we'll have an empty update list. And applying a delete on top of that will result in
+             * ONLY a tombstone in the update list.
+             *
+             * In this case, we should leave the selected update unset to indicate that we want to
+             * keep the same on-disk value but set the stop time pair to indicate that the validity
+             * window ends when this tombstone started.
+             */
+            upd_select->durable_ts = upd_select->start_ts = vpack->start_ts;
+            upd_select->start_txn = vpack->start_txn;
+            /*
+             * Leaving the update unset means that we can skip reconciling. If we've set the stop
+             * time pair because of a tombstone after the on-disk value, we still have work to do so
+             * that is NOT ok. Let's allocate an update equivalent to the on-disk value and continue
+             * on our way!
+             */
+            WT_RET(__wt_scr_alloc(session, 0, &tmp));
+            WT_ERR(__wt_page_cell_data_ref(session, page, vpack, tmp));
+            WT_ERR(__wt_update_alloc(session, tmp, &upd_select->upd, &size, WT_UPDATE_STANDARD));
+            upd_select->upd->ext = 1;
+        }
+        WT_ASSERT(session, upd == NULL || upd_select->upd->type != WT_UPDATE_TOMBSTONE);
 
         /*
          * Finalize the timestamps and transactions, checking if the update is globally visible and
          * nothing needs to be written.
          */
-        if ((upd_select->stop_ts == WT_TS_MAX && upd_select->stop_txn == WT_TXN_MAX) &&
-          ((upd_select->start_ts == WT_TS_NONE && upd_select->start_txn == WT_TXN_NONE) ||
-              __wt_txn_visible_all(session, upd_select->start_txn, upd_select->start_ts))) {
+        if (__wt_txn_visible_all(session, upd_select->start_txn, upd_select->start_ts)) {
```

Comment:
I feel this part of checking and assigning can be completely removed. Already at the beginning the start and stop time pair is assigned properly. Irrespective of the visibility keeping start and stop time pair is better. Otherwise, already set the stop time pair may be reset here.

Reply:
62b594a

I agree with what you're saying. There is no value in fudging the start and stop time pair.